File: benchmarks/benchmark.go
package main

import (
	"fmt"
	"time"
)

// Global black hole to prevent optimization
var BlackHole interface{}

// Global volatile input (set dynamically)
var V_INT = 1

// Measure with loops
func measure(name string, iterations int, f func() interface{}) {
	// Warmup
	for i := 0; i < iterations/10; i++ {
		BlackHole = f()
	}

	start := time.Now()
	for i := 0; i < iterations; i++ {
		BlackHole = f()
	}
	elapsed := time.Since(start)

	// Average time
	totalNs := elapsed.Nanoseconds()
	avgNs := float64(totalNs) / float64(iterations)
	avgUs := avgNs / 1000.0
	avgMs := avgUs / 1000.0

	// Report time per ONE operation (avg)
	if avgUs < 1.0 {
		fmt.Printf("%s\nSure: %.4f ns\n", name, avgNs)
		fmt.Printf("Sure: %.6f µs\n", avgUs) 
	} else if avgMs < 1.0 {
		fmt.Printf("%s\nSure: %.4f µs\n", name, avgUs)
	} else {
		fmt.Printf("%s\nSure: %.4f ms\n", name, avgMs)
	}
	fmt.Println("--------------------------------")
}

func main() {
	// Initialize V_INT dynamically based on time to ensure it's not a compile-time constant
	if time.Now().Year() > 2000 {
		V_INT = 1
	} else {
		V_INT = 2
	}
	
	fmt.Println("Go Benchmark Basliyor...")
	fmt.Printf("Volatile Input Scale: %d\n", V_INT)
	fmt.Println("--------------------------------")

	v30 := 30 * V_INT
	measure("Test 1: Fibonacci(30)", 100, func() interface{} {
		return fibonacci(v30)
	})

	v20 := int64(20 * V_INT)
	measure("Test 2: Factorial(20)", 100000, func() interface{} {
		return factorial(v20)
	})

	v3 := 3 * V_INT
	v8 := 8 * V_INT
	measure("Test 3: Ackermann(3, 8)", 50, func() interface{} {
		return ackermann(v3, v8)
	})

	v18 := 18 * V_INT
	v12 := 12 * V_INT
	v6 := 6 * V_INT
	measure("Test 4: Tak(18, 12, 6)", 2000, func() interface{} {
		return tak(v18, v12, v6)
	})

	v1M := 1000000 * V_INT
	measure("Test 5: Loop 1M", 10000, func() interface{} {
		return loopTest(v1M)
	})

	v10K := 10000 * V_INT
	measure("Test 6: Sieve(10K)", 50000, func() interface{} {
		return sieve(v10K)
	})

	v1K := 1000 * V_INT
	measure("Test 7: BubbleSort(1K)", 20000, func() interface{} {
		return bubbleSort(v1K)
	})

	measure("Test 8: StringConcat(1K)", 100000, func() interface{} {
		return stringConcat(v1K)
	})

	measure("Test 9: ArrayMemory(10K)", 100000, func() interface{} {
		return arrayMemory(v10K)
	})

	measure("Test 10: StringAlloc(1K)", 10000, func() interface{} {
		return stringAlloc(v1K)
	})

	measure("Test 11: JSONBuild(1K)", 10000, func() interface{} {
		return jsonBuild(v1K)
	})
	
	// Force usage of BlackHole
	if BlackHole == nil {
		fmt.Print("")
	}
}

// 1. Fibonacci
func fibonacci(n int) int {
	if n <= 1 {
		return n
	}
	return fibonacci(n-1) + fibonacci(n-2)
}

// 2. Factorial
func factorial(n int64) int64 {
	if n <= 1 {
		return 1
	}
	return n * factorial(n-1)
}

// 3. Ackermann
func ackermann(m, n int) int {
	if m == 0 {
		return n + 1
	}
	if n == 0 {
		return ackermann(m-1, 1)
	}
	return ackermann(m-1, ackermann(m, n-1))
}

// 4. Tak
func tak(x, y, z int) int {
	if y < x {
		return tak(tak(x-1, y, z), tak(y-1, z, x), tak(z-1, x, y))
	}
	return z
}

// 5. Loop used for summation (Loop 1M)
func loopTest(iterations int) int64 {
	var sum int64 = 0
	inc := int64(V_INT)
	if inc == 0 { inc = 1 } // Ensure non-zero
	for i := 0; i < iterations; i++ {
		// Xorshift mix
		x := int64(i)
		x ^= x << 13
		x ^= x >> 17
		x ^= x << 5
		sum += x + inc
	}
	return sum
}

// 6. Sieve
func sieve(n int) int {
	flags := make([]bool, n+1)
	count := 0
	for i := 2; i <= n; i++ {
		if !flags[i] {
			count++
			for k := i * i; k <= n; k += i {
				flags[k] = true
			}
		}
	}
	return count
}

// 7. Bubble Sort
func bubbleSort(size int) int64 {
	arr := make([]int, size)
	// Use runtime V_INT to prevent compile-time generation
	seed := 12345 * V_INT 
	for i := 0; i < size; i++ {
		seed = (seed * 1103515245 + 12345) % 32768
		arr[i] = seed
	}

	vZero := V_INT ^ 1 // 0
	for i := 0; i < size; i++ {
		for j := 0; j < size-1; j++ {
			// runtime check
			if arr[j] > arr[j+1] + vZero {
				arr[j], arr[j+1] = arr[j+1], arr[j]
			}
		}
	}

	var sum int64 = 0
	for _, v := range arr {
		sum += int64(v)
	}
	return sum
}

// 8. String Concat
func stringConcat(iterations int) int {
	result := ""
	for i := 0; i < iterations; i++ {
		result = result + "a"
	}
	return len(result)
}

// 9. Array Memory
func arrayMemory(size int) int {
	arr := make([]int, size)
	for i := 0; i < size; i++ {
		arr[i] = i
	}
	sum := 0
	for i := 0; i < size; i++ {
		sum = sum + arr[i]
	}
	return sum
}

// 10. String Alloc
func stringAlloc(iterations int) int {
	result := ""
	for i := 0; i < iterations; i++ {
		result = result + "Merhaba Dunya "
	}
	return len(result)
}

// 11. JSON Build (simulated using map[string]interface{})
type JSONItem struct {
	ID     int
	Name   string
	Value  int
	Active bool
}

func jsonBuild(size int) int {
	data := make([]JSONItem, size)
	for i := 0; i < size; i++ {
		data[i] = JSONItem{
			ID:     i,
			Name:   "Item",
			Value:  i * 10,
			Active: true,
		}
	}
	sum := 0
	for i := 0; i < size; i++ {
		sum += data[i].Value
	}
	return sum
}

File: benchmarks/benchmark.rs
use std::time::Instant;
use std::fmt::Write;

// Helper to prevent optimization
// Helper to prevent optimization
// This ensures that the compiler treats the value as used and unknown
fn black_box<T>(dummy: T) -> T {
    unsafe {
        let ret = std::ptr::read_volatile(&dummy);
        std::mem::forget(dummy);
        ret
    }
}

// Measure with loops for accuracy
fn measure<F>(name: &str, iterations: u32, f: F)
where
    F: Fn(),
{
    // Warmup
    for _ in 0..iterations / 10 {
        f();
    }

    let start = Instant::now();
    for _ in 0..iterations {
        f();
    }
    let duration = start.elapsed();
    
    // Average time per iteration
    let avg_ns = duration.as_nanos() as f64 / iterations as f64;
    let avg_us = avg_ns / 1000.0;
    let avg_ms = avg_us / 1000.0;

    println!("{}", name);
    if avg_us < 1.0 {
        println!("Sure: {:.4} ns", avg_ns); 
        println!("Sure: {:.6} µs", avg_us); 
    } else if avg_ms < 1.0 {
        println!("Sure: {:.4} ms", avg_ms);
        println!("Sure: {:.4} µs", avg_us);
    } else {
        println!("Sure: {:.4} ms", avg_ms);
    }
    println!("--------------------------------");
}

fn main() {
    println!("Rust Benchmark Basliyor...");
    println!("--------------------------------");

    // Arguments passed through black_box are treated as runtime values
    measure("Test 1: Fibonacci(30)", 100, || {
        black_box(fibonacci(black_box(30)));
    });

    measure("Test 2: Factorial(20)", 100000, || {
        black_box(factorial(black_box(20)));
    });

    measure("Test 3: Ackermann(3, 8)", 50, || {
        black_box(ackermann(black_box(3), black_box(8)));
    });

    measure("Test 4: Tak(18, 12, 6)", 200, || {
        black_box(tak(black_box(18), black_box(12), black_box(6)));
    });

    measure("Test 5: Loop 1M", 1000, || {
        black_box(loop_test(black_box(1000000)));
    });

    measure("Test 6: Sieve(10K)", 5000, || {
        black_box(sieve(black_box(10000)));
    });

    measure("Test 7: BubbleSort(1K)", 2000, || {
        black_box(bubble_sort(black_box(1000)));
    });

    measure("Test 8: StringConcat(1K)", 10000, || {
        black_box(string_concat(black_box(1000)));
    });

    measure("Test 9: ArrayMemory(10K)", 10000, || {
        black_box(array_memory(black_box(10000)));
    });

    measure("Test 10: StringAlloc(1K)", 10000, || {
        black_box(string_alloc(black_box(1000)));
    });

    measure("Test 11: JSONBuild(1K)", 10000, || {
        black_box(json_build(black_box(1000)));
    });
}

// 1. Fibonacci
fn fibonacci(n: i32) -> i32 {
    if n <= 1 {
        return n;
    }
    fibonacci(n - 1) + fibonacci(n - 2)
}

// 2. Factorial
fn factorial(n: i64) -> i64 {
    if n <= 1 {
        return 1;
    }
    n * factorial(n - 1)
}

// 3. Ackermann
fn ackermann(m: i32, n: i32) -> i32 {
    if m == 0 {
        return n + 1;
    }
    if n == 0 {
        return ackermann(m - 1, 1);
    }
    ackermann(m - 1, ackermann(m, n - 1))
}

// 4. Tak
fn tak(x: i32, y: i32, z: i32) -> i32 {
    if y < x {
        return tak(tak(x - 1, y, z), tak(y - 1, z, x), tak(z - 1, x, y));
    }
    z
}

// 5. Loop 1M
fn loop_test(iterations: i64) -> i64 {
    let mut sum: i64 = 0;
    let inc = black_box(1); // Force runtime dependency
    for i in 0..iterations {
        // Xorshift mix
        let mut x = i;
        x ^= x << 13;
        x ^= x >> 17;
        x ^= x << 5;
        sum = sum.wrapping_add(x).wrapping_add(inc);
    }
    sum
}

// 6. Sieve
fn sieve(n: usize) -> usize {
    let mut flags = vec![false; n + 1];
    let mut count = 0;
    for i in 2..=n {
        if !flags[i] {
            count += 1;
            let mut k = i * i;
            while k <= n {
                flags[k] = true;
                k += i;
            }
        }
    }
    count
}

// 7. Bubble Sort
fn bubble_sort(size: usize) -> i64 {
    let mut arr = vec![0; size];
    // Use runtime black_box to prevent compile-time generation
    let mut seed: i64 = 12345 * black_box(1); 
    for i in 0..size {
        seed = (seed * 1103515245 + 12345) % 32768;
        arr[i] = seed;
    }

    let v_zero = black_box(0);
    for _ in 0..size {
        for j in 0..size - 1 {
            // Force runtime comparison
            if arr[j] > arr[j + 1] + v_zero {
                arr.swap(j, j + 1);
            }
        }
    }
    
    arr.iter().sum()
}

// 8. String Concat
fn string_concat(iterations: usize) -> usize {
    let mut result = String::new();
    for _ in 0..iterations {
        result.push_str("a"); // or push('a')
    }
    result.len()
}

// 9. Array Memory
fn array_memory(size: usize) -> usize {
    let mut arr = Vec::with_capacity(size);
    for i in 0..size {
        arr.push(i);
    }
    let mut sum = 0;
    for i in 0..size {
        sum += arr[i];
    }
    sum
}

// 10. String Alloc
fn string_alloc(iterations: usize) -> usize {
    let mut result = String::new();
    for _ in 0..iterations {
        result.push_str("Merhaba Dunya ");
    }
    result.len()
}

// 11. JSON Build (Struct emulation)
struct JsonItem {
    id: usize,
    name: String,
    value: usize,
    active: bool,
}

fn json_build(size: usize) -> usize {
    let mut data = Vec::with_capacity(size);
    for i in 0..size {
        data.push(JsonItem {
            id: i,
            name: "Item".to_string(),
            value: i * 10,
            active: true,
        });
    }
    
    let mut sum = 0;
    for item in &data {
        sum += item.value;
    }
    sum
}
